logging/redaction/log_redaction_final.py

```diff
@@ -0,0 +1,218 @@
+        return {'values': [{'string_value': payload}]}
+
+    def setup(self):
+        '''Initialize DLP client'''
+        if self.dlp_client:
+            return
+        # self.dlp_client = dlp_v2.DlpServiceClient(client_options={'quota_project_id', self.project_id})
+        self.dlp_client = dlp_v2.DlpServiceClient()
+        if not self.dlp_client:
+            logging.error('Cannot create Google DLP Client')
+            raise PipelineError('Cannot create Google DLP Client')
+
+    def process(self, logs):
+        # Construct the `table`. For more details on the table schema, please see
+        # https://cloud.google.com/dlp/docs/reference/rest/v2/ContentItem#Table
+        table = {
+            'table': {
+                'headers': [{'name': 'textPayload'}],
+                'rows': map(self._log_to_row, logs)
+            }
+        }
+
+        response = self.dlp_client.deidentify_content(
+            request={
+                'parent': f'projects/{self.project_id}',
+                'inspect_config': INSPECT_CFG,
+                'deidentify_config': REDACTION_CFG,
+                'item': table,
+            })
+
+        # replace payload with redacted version
+        modified_logs = []
+        for i, log in enumerate(logs):
+            log['textPayload'] = response.item.table.rows[i].values[0].string_value
+            # you may consider changing insert ID if the project already has a copy
+            # of this log (e.g. log['insertId'] = 'deid-' + log['insertId'])
+            # For more details about insert ID, please see:
+            # https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#FIELDS.insert_id
+            modified_logs.append(log)
+        yield modified_logs
+
+
+class IngestLogs(DoFn):
+    '''Ingest payloads into destination log'''
+
+    def __init__(self, destination_log_name):
+        self.destination_log_name = destination_log_name
+        self.logger = None
+
+    def _replace_log_name(self, entry):
+        # update log name in the entry with destination log
+        entry['logName'] = self.logger.name
+        return entry
+
+    def setup(self):
+        # initialize logging client
+        if self.logger:
+            return
+
+        logging_client = logging_v2.Client()
+        if not logging_client:
+            logging.error('Cannot create Google Logging Client')
+            raise PipelineError('Cannot create Google Logging Client')
+        self.logger = logging_client.logger(self.destination_log_name)
+        if not self.logger:
+            logging.error('Google client library cannot create Logger object')
+            raise PipelineError('Google client library cannot create Logger object')
+
+    def process(self, element):
+        if self.logger:
+            logs = list(map(self._replace_log_name, element))
+            self.logger.client.logging_api.write_entries(logs)
+        yield logs
+
+
+def run(pubsub_subscription: str,
```

Comment:
Would it be helpful to add region tags here? I know this code is very specific to pubsub and beam, but I think it could be useful to have one for processing logs in a beam pipeline, and maybe one for redacting logs.

Reply:
There is no use for this as a code snippet at the moment. I do not want to create an orphan code sample.